File: C-2/cash.c
// Greedy algorithms
#include <stdio.h>
// round function in math header files
#include <math.h>
#include <cs50.h>

// four function or senarios to achieve the Greedy Algorithms
// a number must be postitve which is the first function.Notice at the final line of code. I covert the dollar into s using round fucntion
// the seconde is a senario when the number is bigger than 5, in this case (it could be bigger than 5 or smaller than 5)
// the third is a senario when the number is bigger than 10. if is it we count the time that it - 10 then return the counter times d
// the forth is a senario where the number is bigger than 25, in this case we count the time that it - 23 until it can't and then return the counter times q
int get_positive_float(void);
int s_Less_than_10(int s);
int s_less_than_25(int s);
int s_bigger_than_25(int s);

int main(void)
{
    // using the frist funcion to get a positive number and a round-digits-converted number  0.1 ---into---10
    //pormpt the quesiton
    int s = get_positive_float();
    //frist senario when s > 25
    if (s >= 25)
    {
        // q is the number of qurater we need to give
        int q = s_bigger_than_25(s);
        // s needs to change since we already minus 25 q times
        s = s - 25 * q;

        //the second senario where number is bigger than ten.
        if (s >= 10)
        {

            int d = s_less_than_25(s);
            //  s needs to change again since we already minus 10 d times
            s = s - d * 10 ;

            //the last senario where the number is bigger or less than 5.
            int p = s_Less_than_10(s);

            //Notice the printf has three vairbles: p d q , that is because we need to add the times we minus 25-q 10-d 5-c and 1-p
            // p in this case has two posibilities 1. p is the time the number minus 5  2. p is the time the number minus 5 and the time the number minus 1
            printf("%i\n", p + d + q);
        }
        //the last senario where the number is bigger or less than 5.
        else
        {
            int p = s_Less_than_10(s);
            printf("%i\n", q + p);
        }

    }
    else if (s >= 10)
    {
        int d = s_less_than_25(s);
        s = s - d * 10 ;
        int p = s_Less_than_10(s);

        printf("%i\n", d + p);
    }
    else
    {
        int p = s_Less_than_10(s);

        printf("%i\n", p);
    }
}


// only give a output that is a converted positive vaule
int get_positive_float(void)
{
    float dollar;
    do
    {
        dollar = get_float("Change Owed: ");
    }
    while (dollar < 0);
    int n = round(dollar * 100);
    return n;
}

// Not only needs a input but also a output
//when s is less than 10 in which it has two possibilities
int s_Less_than_10(int s)
{
    if (s >= 5)
    {
        int c = 0;
        do
        {
            s = s - 5;
            c = c + 1;
        }
        while (s >= 5);
        //when s < 5 , it can only be give by a penney
        //Ergo the number of penneies when need is equal to s
        //P is the total number of coins 5 and 1 needed to give to the coustomer
        int p = s + c;
        return p;
    }
    else
    {
        int p = s;
        return p;
    }
}

// when is less than 25 but bigger than 10.
int s_less_than_25(int s)
{
    int d = 0;
    do
    {
        s = s - 10;
        d = d + 1;
    }
    while (s >= 10);
    return d;
}

int s_bigger_than_25(int s)
{
    int q = 0;
    do
    {
        s = s - 25;
        q = q + 1;
    }
    while (s >= 25);
    return q;
}

File: pset6/pset6_part1/cash.py
import cs50


def main():
    changes = get_positive_float()
    # Notice we need to convert it into int
    total_times = int(cash(changes))

    print(total_times)


def get_positive_float():
    # here we use the forever loop
    while True:
        changes = cs50.get_float("Changes Owed: ")
        # until the user put in this range
        # we will keep asking them
        if changes > 0:
            # break out of the loop and then go to the next line
            # that is return n
            break
    return changes


def cash(changes):
    money = changes * 100
    if money > 25:
        q = condition(money, 25)
        money = money - q * 25
        d = condition(money, 10)
        money = money - d * 10
        n = condition(money, 5)
        money = money - n * 5
        c = condition(money, 1)
        return q + d + n + c
    elif money > 10:
        d = condition(money, 10)
        print(d)
        money = money - d * 10
        n = condition(money, 5)
        money = money - n * 5
        c = condition(money, 1)
        return d + n + c
    elif money > 5:
        n = condition(money, 5)
        money = money - n * 5
        c = condition(money, 1)
        return n + c
    else:
        return money


# this function take money and the coins_value as input
# first it will compare the two
# and find how much a type of coins we need to reduce the money as much as possible (Greedy)
def condition(money, n):
    if money >= n:
        remain = money % n
        # the times is n * times = (money - remain)
        times = (money - remain) / n
    else:
        # if the coin is bigger than the money
        # then we need zero number of this type coin
        times = 0
    return times


main()